cmd/record/record.go

```diff
@@ -0,0 +1,217 @@
+// Copyright 2021 Authors of Hubble
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package record
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"os"
+	"os/signal"
+	"strconv"
+	"strings"
+	"syscall"
+	"time"
+
+	recorderpb "github.com/cilium/cilium/api/v1/recorder"
+	"github.com/cilium/hubble/cmd/common/config"
+	"github.com/cilium/hubble/cmd/common/conn"
+	"github.com/cilium/hubble/cmd/common/template"
+	"github.com/spf13/cobra"
+	"github.com/spf13/pflag"
+	"github.com/spf13/viper"
+	"google.golang.org/grpc"
+)
+
+const (
+	fitlerSep     = " "
+	resetLastLine = "\033[A\033[2K"
+)
+
+var fileSinkPrefix string
+var maxCaptureLength uint32
+
+// New creates a new record subcommand
+func New(vp *viper.Viper) *cobra.Command {
+	recordCmd := &cobra.Command{
+		Use:   "record [flags] filter1 filter2 ... filterN",
+		Short: "Capture and record network packets",
+		Long:  "Capture and record network packets into a pcap file stored on the Hubble server",
+		Example: `
+Record all TCP traffic from 192.168.1.0/24 (any source port) to 10.0.0.0/16 port 80:
+
+  hubble record "192.168.1.0/24 0 10.0.0.0/16 80 TCP"
+
+Multiple filters may be specified, in which case any matching packet will be recorded.
+The filter syntax is "srcPrefix srcPort dstPrefix dstPort proto". Currently supported
+protocols are TCP, UDP and ANY.`,
+		Hidden: true, // this command is experimental
+		RunE: func(cmd *cobra.Command, args []string) error {
+			filters, err := parseFilters(args)
+			if err != nil {
+				return err
+			}
+
+			ctx, cancel := context.WithCancel(context.Background())
+			defer cancel()
+
+			hubbleConn, err := conn.New(ctx, vp.GetString(config.KeyServer), vp.GetDuration(config.KeyTimeout))
+			if err != nil {
+				return err
+			}
+			defer hubbleConn.Close()
+
+			return runRecord(ctx, hubbleConn, filters)
+		},
+	}
+
+	recorderFlags := pflag.NewFlagSet("recorder", pflag.ContinueOnError)
+	recorderFlags.StringVar(&fileSinkPrefix, "file-prefix", "hubble", "File prefix of the resulting pcap file")
+	recorderFlags.Uint32Var(&maxCaptureLength, "max-capture-len", 0, "Sets the maximum capture length (zero for full capture)")
+
+	recordCmd.Flags().AddFlagSet(recorderFlags)
+	recordCmd.SetUsageTemplate(template.Usage(config.ServerFlags, recorderFlags))
+
+	return recordCmd
+}
+
+func parseProto(proto string) (recorderpb.Protocol, error) {
```

Comment:
does proto not generate a function similar to this already?

Reply:
Not to my knowledge.